spec/UrlMiddleware.coffee

```diff
@@ -0,0 +1,136 @@
+noflo = require 'noflo'
+
+unless noflo.isBrowser()
+  chai = require 'chai'
+  path = require 'path'
+  baseDir = path.resolve __dirname, '../'
+else
+  baseDir = 'noflo-ui'
+
+describe 'URL Middleware', ->
+  c = null
+  actionIn = null
+  passAction = null
+  newAction = null
+  before (done) ->
+    @timeout 4000
+    loader = new noflo.ComponentLoader baseDir
+    loader.load 'ui/UrlMiddleware', (err, instance) ->
+      return done err if err
+      c = instance
+      actionIn = noflo.internalSocket.createSocket()
+      c.inPorts.action.attach actionIn
+      actionIn.port = 'action'
+      c.start()
+      c.network.once 'start', ->
+        done()
+  beforeEach ->
+    passAction = noflo.internalSocket.createSocket()
+    c.outPorts.pass.attach passAction
+    passAction.port = 'pass'
+    newAction = noflo.internalSocket.createSocket()
+    c.outPorts.action.attach newAction
+    newAction.port = 'action'
+  afterEach ->
+    c.outPorts.pass.detach passAction
+    c.outPorts.action.detach newAction
+  after ->
+    window.location.hash = ''
+
+  send = (socket, action, payload) ->
+    actionParts = action.split ':'
+    socket.beginGroup part for part in actionParts
+    socket.send payload
+    socket.endGroup part for part in actionParts
+    
+  receive = (socket, expected, check, done) ->
+    received = []
+    onBeginGroup = (group) ->
+      received.push "< #{group}"
+    onData = (data) ->
+      received.push 'DATA'
+      check data
+    onEndGroup = (group) ->
+      received.push "> #{group}"
+      return unless received.length >= expected.length
+      socket.removeListener 'begingroup', onBeginGroup
+      socket.removeListener 'data', onData
+      socket.removeListener 'endgroup', onEndGroup
+      chai.expect(received).to.eql expected
+      done()
+    socket.on 'begingroup', onBeginGroup
+    socket.on 'data', onData
+    socket.on 'endgroup', onEndGroup
+
+  receiveAction = (socket, action, check, done) ->
+    expected = []
+    actionParts = action.split ':'
+    expected.push "< #{part}" for part in actionParts
+    expected.push 'DATA'
+    actionParts.reverse()
+    expected.push "> #{part}" for part in actionParts
+    receive socket, expected, check, done
+
+  receivePass = (socket, action, payload, done) ->
+    check = (data) ->
+      # Strict equality check for passed packets
+      chai.expect(data).to.equal payload
+    receiveAction socket, action, check, done
+
+  describe 'receiving a runtime:connect action', ->
+    it 'should pass it out as-is', (done) ->
+      action = 'runtime:connect'
+      payload =
+        hello: 'world'
```

Comment:
I wanted a test of a passed packet. The contents don't really matter as long as:

* the action is one that the middleware isn't supposed to handle
* we verify the packet gets passed through without manipulation (strict equality check)